File: frontend/lib/i18n/locale.ts
// Locale primitives shared by the server layout and the client provider; no React imports allowed here.

export type Locale = "zh" | "en";

export const LOCALE_COOKIE = "lgtm-locale";
export const DEFAULT_LOCALE: Locale = "zh";

export function isLocale(value: unknown): value is Locale {
  return value === "zh" || value === "en";
}

// negotiate returns the highest-q supported tag from an Accept-Language header, or DEFAULT_LOCALE.
export function negotiate(acceptLanguage: string | null | undefined): Locale {
  if (!acceptLanguage) return DEFAULT_LOCALE;

  const tags = acceptLanguage
    .split(",")
    .map((part) => {
      const [rawTag, ...params] = part.trim().split(";");
      const qParam = params.map((p) => p.trim()).find((p) => p.startsWith("q="));
      const parsed = qParam ? Number.parseFloat(qParam.slice(2)) : 1;
      // A malformed q is treated as lowest priority rather than an error.
      return { tag: rawTag.trim().toLowerCase(), weight: Number.isNaN(parsed) ? 0 : parsed };
    })
    .filter((t) => t.tag !== "")
    .sort((a, b) => b.weight - a.weight);

  for (const { tag } of tags) {
    if (tag === "zh" || tag.startsWith("zh-")) return "zh";
    if (tag === "en" || tag.startsWith("en-")) return "en";
  }
  return DEFAULT_LOCALE;
}

File: backend/internal/i18n/locale.go
// Package i18n resolves the review output locale. It is the Go counterpart of frontend/lib/i18n/locale.ts.
// Naming note: "locale" means UI / review language here. The unrelated notion of a PR's programming
// language lives in internal/api/lang.go.
package i18n

import (
	"sort"
	"strconv"
	"strings"
)

type Locale string

const (
	ZH Locale = "zh"
	EN Locale = "en"
)

// Normalize maps arbitrary input onto a supported locale, falling back to def rather than erroring.
func Normalize(v string, def Locale) Locale {
	switch Locale(strings.ToLower(strings.TrimSpace(v))) {
	case ZH:
		return ZH
	case EN:
		return EN
	default:
		return def
	}
}

// FromAcceptLanguage returns the highest-q supported tag in the header, or def.
func FromAcceptLanguage(header string, def Locale) Locale {
	if strings.TrimSpace(header) == "" {
		return def
	}

	type weighted struct {
		tag    string
		weight float64
	}
	var tags []weighted
	for _, part := range strings.Split(header, ",") {
		fields := strings.Split(strings.TrimSpace(part), ";")
		tag := strings.ToLower(strings.TrimSpace(fields[0]))
		if tag == "" {
			continue
		}
		weight := 1.0
		for _, p := range fields[1:] {
			p = strings.TrimSpace(p)
			if !strings.HasPrefix(p, "q=") {
				continue
			}
			// A malformed q is treated as lowest priority rather than an error.
			if q, err := strconv.ParseFloat(strings.TrimPrefix(p, "q="), 64); err == nil {
				weight = q
			} else {
				weight = 0
			}
		}
		tags = append(tags, weighted{tag: tag, weight: weight})
	}
	sort.SliceStable(tags, func(i, j int) bool { return tags[i].weight > tags[j].weight })

	for _, t := range tags {
		switch {
		case t.tag == "zh" || strings.HasPrefix(t.tag, "zh-"):
			return ZH
		case t.tag == "en" || strings.HasPrefix(t.tag, "en-"):
			return EN
		}
	}
	return def
}
